### PR TITLE
Minimise EntityFertilizeEggEvent and add sniffer

### PR DESCRIPTION
**Pull request by lynxplay**
Minimises the EntityFertilizeEggEvent by moving the duplicate code out into the CraftEventFactory utility.
Additionally, this adds sniffers as a fertilising entity by calling the event for them breeding and producing the sniffer egg item.

### CODE REVIEW

File: patches/server/0930-Add-EntityFertilizeEggEvent.patch [hunk 1]
```diff
@@ -5,37 +5,100 @@ Subject: [PATCH] Add EntityFertilizeEggEvent
...
++     * If the event is cancelled, this method also resets the love on both the {@code breeding} and {@code other} entity.
++     *
++     * @param breeding the entity on which #spawnChildFromBreeding was called.
++     * @param other    the partner of the entity.
++     * @return the event after it was called. The instance may be used to retrieve the experience of the event.
++     */
++    public static io.papermc.paper.event.entity.EntityFertilizeEggEvent callEntityFertilizeEggEvent(net.minecraft.world.entity.animal.Animal breeding,
++                                                                                                    net.minecraft.world.entity.animal.Animal other) {
++        net.minecraft.server.level.ServerPlayer serverPlayer = breeding.getLoveCause();
++        if (serverPlayer == null) other.getLoveCause();
```
Comment (Lulu13022002) [inline patches/server/0930-Add-EntityFertilizeEggEvent.patch]: Is this intended ?

File: patches/server/0930-Add-EntityFertilizeEggEvent.patch [hunk 2]
```diff
@@ -5,37 +5,100 @@ Subject: [PATCH] Add EntityFertilizeEggEvent
...
++        final int experience = breeding.getRandom().nextInt(7) + 1; // From Animal#spawnChildFromBreeding(ServerLevel, Animal)
++
++        final var event = new io.papermc.paper.event.entity.EntityFertilizeEggEvent((org.bukkit.entity.LivingEntity) breeding.getBukkitEntity(), (org.bukkit.entity.LivingEntity) other.getBukkitEntity(), serverPlayer == null ? null : serverPlayer.getBukkitEntity(), breeding.breedItem == null ? null : org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(breeding.breedItem).clone(), experience);
```
Comment (Lulu13022002) [inline patches/server/0930-Add-EntityFertilizeEggEvent.patch]: You can use CraftItemStack#asCraftCopy

**Review by Lulu13022002 — COMMENTED** [review]
Can you rewrite some javadoc of the api event, i was lazy and there's some stuff that was badly copied from the breed event. Also the javadoc on the CraftEventFactory is really clean but since it's the only method i'm not sure about that

File: patches/server/0930-Add-EntityFertilizeEggEvent.patch (hunk 1 above)
Comment (lynxplay) [reply patches/server/0930-Add-EntityFertilizeEggEvent.patch]: No it is not xD

File: patches/server/0930-Add-EntityFertilizeEggEvent.patch (hunk 2 above)
Comment (lynxplay) [reply patches/server/0930-Add-EntityFertilizeEggEvent.patch]: asCraftCopy takes a bukkit item stack. I am confused what you mean here.

Comment (Lulu13022002) [reply patches/server/0930-Add-EntityFertilizeEggEvent.patch]: oh nvm

**Review by Lulu13022002 — APPROVED** [review]
Thanks :heart:

**Review by Machine-Maker — APPROVED** [review]

File: patches/server/0939-Add-EntityFertilizeEggEvent.patch [hunk 3]
```diff
@@ -5,37 +5,100 @@ Subject: [PATCH] Add EntityFertilizeEggEvent
...
++     * @param other    the partner of the entity.
++     * @return the event after it was called. The instance may be used to retrieve the experience of the event.
++     */
++    public static io.papermc.paper.event.entity.EntityFertilizeEggEvent callEntityFertilizeEggEvent(net.minecraft.world.entity.animal.Animal breeding,
++                                                                                                    net.minecraft.world.entity.animal.Animal other) {
++        net.minecraft.server.level.ServerPlayer serverPlayer = breeding.getLoveCause();
++        if (serverPlayer == null) serverPlayer = other.getLoveCause();
++        final int experience = breeding.getRandom().nextInt(7) + 1; // From Animal#spawnChildFromBreeding(ServerLevel, Animal)
++
++        final var event = new io.papermc.paper.event.entity.EntityFertilizeEggEvent((org.bukkit.entity.LivingEntity) breeding.getBukkitEntity(), (org.bukkit.entity.LivingEntity) other.getBukkitEntity(), serverPlayer == null ? null : serverPlayer.getBukkitEntity(), breeding.breedItem == null ? null : org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(breeding.breedItem).clone(), experience);
```
Comment (kennytv) [inline patches/server/0939-Add-EntityFertilizeEggEvent.patch]: .

**Review by kennytv — CHANGES_REQUESTED** [review]

**Review by kennytv — APPROVED** [review]

**Pull request merged**